# Model Fit in Linear Regression - Lab

## Introduction
In this lab, you'll learn how to evaluate your model results and you'll learn how to select the appropriate features using stepwise selection.

## Objectives
You will be able to:
* Use stepwise selection methods to determine the most important features for a model
* Use recursive feature elimination to determine the most important features for a model

## The Boston Housing Data once more

We pre-processed the Boston Housing data the same way we did before:

- We dropped `'ZN'` and `'NOX'` completely
- We categorized `'RAD'` in 3 bins and `'TAX'` in 4 bins
- We transformed `'RAD'` and `'TAX'` to dummy variables and dropped the first variable
- We used min-max-scaling on `'B'`, `'CRIM'`, and `'DIS'` (and logtransformed all of them first, except `'B'`)
- We used standardization on `'AGE'`, `'INDUS'`, `'LSTAT'`, and `'PTRATIO'` (and logtransformed all of them first, except for `'AGE'`) 

In [1]:
import pandas as pd
import numpy as np
from sklearn.datasets import load_boston
boston = load_boston()

boston_features = pd.DataFrame(boston.data, columns = boston.feature_names)
boston_features = boston_features.drop(['NOX', 'ZN'],axis=1)

# First, create bins for based on the values observed. 3 values will result in 2 bins
bins = [0,6,  24]
bins_rad = pd.cut(boston_features['RAD'], bins)
bins_rad = bins_rad.cat.as_unordered()

# First, create bins for based on the values observed. 4 values will result in 3 bins
bins = [0, 270, 360, 712]
bins_tax = pd.cut(boston_features['TAX'], bins)
bins_tax = bins_tax.cat.as_unordered()

tax_dummy = pd.get_dummies(bins_tax, prefix='TAX', drop_first=True)
rad_dummy = pd.get_dummies(bins_rad, prefix='RAD', drop_first=True)
boston_features = boston_features.drop(['RAD', 'TAX'], axis=1)
boston_features = pd.concat([boston_features, rad_dummy, tax_dummy], axis=1)

age = boston_features['AGE']
b = boston_features['B']
logcrim = np.log(boston_features['CRIM'])
logdis = np.log(boston_features['DIS'])
logindus = np.log(boston_features['INDUS'])
loglstat = np.log(boston_features['LSTAT'])
logptratio = np.log(boston_features['PTRATIO'])

# Min-Max scaling
boston_features['B'] = (b-min(b))/(max(b)-min(b))
boston_features['CRIM'] = (logcrim-min(logcrim))/(max(logcrim)-min(logcrim))
boston_features['DIS'] = (logdis-min(logdis))/(max(logdis)-min(logdis))

# Standardization
boston_features['AGE'] = (age-np.mean(age))/np.sqrt(np.var(age))
boston_features['INDUS'] = (logindus-np.mean(logindus))/np.sqrt(np.var(logindus))
boston_features['LSTAT'] = (loglstat-np.mean(loglstat))/np.sqrt(np.var(loglstat))
boston_features['PTRATIO'] = (logptratio-np.mean(logptratio))/(np.sqrt(np.var(logptratio)))

## Perform stepwise selection

The function for stepwise selection is copied below. Use this function provided on your preprocessed Boston Housing data.

In [2]:
import statsmodels.api as sm

def stepwise_selection(X, y, 
                       initial_list=[], 
                       threshold_in=0.01, 
                       threshold_out = 0.05, 
                       verbose=True):
    """ Perform a forward-backward feature selection 
    based on p-value from statsmodels.api.OLS
    Arguments:
        X - pandas.DataFrame with candidate features
        y - list-like with the target
        initial_list - list of features to start with (column names of X)
        threshold_in - include a feature if its p-value < threshold_in
        threshold_out - exclude a feature if its p-value > threshold_out
        verbose - whether to print the sequence of inclusions and exclusions
    Returns: list of selected features 
    Always set threshold_in < threshold_out to avoid infinite looping.
    See https://en.wikipedia.org/wiki/Stepwise_regression for the details
    """
    included = list(initial_list)
    while True:
        changed=False
        # forward step
        excluded = list(set(X.columns)-set(included))
        new_pval = pd.Series(index=excluded)
        for new_column in excluded:
            model = sm.OLS(y, sm.add_constant(pd.DataFrame(X[included+[new_column]]))).fit()
            new_pval[new_column] = model.pvalues[new_column]
        best_pval = new_pval.min()
        if best_pval < threshold_in:
            best_feature = new_pval.idxmin()
            included.append(best_feature)
            changed=True
            if verbose:
                print('Add  {:30} with p-value {:.6}'.format(best_feature, best_pval))

        # backward step
        model = sm.OLS(y, sm.add_constant(pd.DataFrame(X[included]))).fit()
        # use all coefs except intercept
        pvalues = model.pvalues.iloc[1:]
        worst_pval = pvalues.max() # null if pvalues is empty
        if worst_pval > threshold_out:
            changed=True
            worst_feature = pvalues.argmax()
            included.remove(worst_feature)
            if verbose:
                print('Drop {:30} with p-value {:.6}'.format(worst_feature, worst_pval))
        if not changed:
            break
    return included

In [3]:
boston_features.head()

CRIM     INDUS  CHAS     RM       AGE       DIS   PTRATIO         B  \
0  0.000000 -1.704344   0.0  6.575 -0.120013  0.542096 -1.443977  1.000000   
1  0.153211 -0.263239   0.0  6.421  0.367166  0.623954 -0.230278  1.000000   
2  0.153134 -0.263239   0.0  7.185 -0.265812  0.623954 -0.230278  0.989737   
3  0.171005 -1.778965   0.0  6.998 -0.809889  0.707895  0.165279  0.994276   
4  0.250315 -1.778965   0.0  7.147 -0.511180  0.707895  0.165279  1.000000   

      LSTAT  RAD_(6, 24]  TAX_(270, 360]  TAX_(360, 712]  
0 -1.275260            0               1               0  
1 -0.263711            0               0               0  
2 -1.627858            0               0               0  
3 -2.153192            0               0               0  
4 -1.162114            0               0               0

In [4]:
# Your code here
X = boston_features
y = pd.DataFrame(boston.target, columns = ['price'])

result = stepwise_selection(X, y, verbose = True)
print('Resulting Features:')
print(result)

C:\Users\anupj\Anaconda3\envs\learn-env\lib\site-packages\numpy\core\fromnumeric.py:2495: FutureWarning: Method .ptp is deprecated and will be removed in a future version. Use numpy.ptp instead.
  return ptp(axis=axis, out=out, **kwargs)


Add  LSTAT                          with p-value 9.27989e-122
Add  RM                             with p-value 1.98621e-16
Add  PTRATIO                        with p-value 2.5977e-12
Add  DIS                            with p-value 2.85496e-09
Add  B                              with p-value 2.77572e-06
Add  INDUS                          with p-value 0.0017767
Add  CHAS                           with p-value 0.0004737
Resulting Features:
['LSTAT', 'RM', 'PTRATIO', 'DIS', 'B', 'INDUS', 'CHAS']


### Build the final model again in Statsmodels

In [11]:
# Your code here
import statsmodels.api as sm

X_fin = X[result]
X_with_intercept = sm.add_constant(X_fin)
model = sm.OLS(y, X_with_intercept).fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.773
Model:                            OLS   Adj. R-squared:                  0.770
Method:                 Least Squares   F-statistic:                     242.7
Date:                Thu, 19 Dec 2019   Prob (F-statistic):          4.89e-156
Time:                        09:13:33   Log-Likelihood:                -1464.7
No. Observations:                 506   AIC:                             2945.
Df Residuals:                     498   BIC:                             2979.
Df Model:                           7                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          5.0123      2.829      1.772      0.077      -0.545      10.570
LSTAT         -5.6444      0.320    -17.629      0.000      -6.274      -5.015
RM             2.8712      0.388      7.405      0.000       2.109       3.633
PTRATIO       -1.3564      0.227     -5.983      0.000      -1.802      -0.911
DIS           -9.7229      1.326     -7.333      0.000     -12.328      -7.118
B              4.0619      0.934      4.347      0.000       2.226       5.898
INDUS         -1.2099      0.334     -3.619      0.000      -1.867      -0.553
CHAS           2.7988      0.795      3.519      0.000       1.236       4.362
==============================================================================
Omnibus:                      105.185   Durbin-Watson:                   1.099
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              423.621
Skew:                           0.878   Prob(JB):                     1.03e-92
Kurtosis:                       7.124   Cond. No.                         96.7
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

## Trying it with Scikit Learn

In [6]:
from sklearn.linear_model import LinearRegression

scikit_model = LinearRegression()
scikit_model.fit(X_fin,y)
scikit_model.score(X_fin,y)

0.7733354280915333

In [8]:
scikit_model.coef_

array([[-5.64444757,  2.87120814, -1.35644489, -9.72288018,  4.06190831,
        -1.20986342,  2.79880718]])

In [9]:
scikit_model.intercept_

array([5.0122525])

The stepwise procedure mentions that `'INDUS'` was added with a p-value of 0.0017767, but our statsmodels output returns a p-value of 0.000. Use some of the stepwise procedure logic to find the intuition behind this!

In [12]:
scikit_model.get_params()

{'copy_X': True, 'fit_intercept': True, 'n_jobs': None, 'normalize': False}

## Use Feature ranking with recursive feature elimination

Use feature ranking to select the 5 most important features

In [30]:
# Your code here
from sklearn.feature_selection import RFE

scikit_model = LinearRegression()
selector = RFE(scikit_model, n_features_to_select= 5)

selector = selector.fit(X, y.values.ravel())
selector.support_

array([False, False,  True,  True, False,  True, False,  True,  True,
       False, False, False])

In [31]:
selected_cols = X.columns[selector.support_]
selected_cols

Index(['CHAS', 'RM', 'DIS', 'B', 'LSTAT'], dtype='object')

Fit the linear regression model again using the 5 selected columns

In [32]:
# Your code here
scikit_model.fit(X[selected_cols], y)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [35]:
scikit_model.score(X[selected_cols], y)

0.7429807743129864

Now, predict $\hat y$ using your model. You can use `.predict()` in scikit-learn. 

In [37]:
# Your code here
yhat = scikit_model.predict(X[selected_cols])

Now, using the formulas of R-squared and adjusted R-squared below, and your Python/numpy knowledge, compute them and contrast them with the R-squared and adjusted R-squared in your statsmodels output using stepwise selection. Which of the two models would you prefer?

$SS_{residual} = \sum (y - \hat{y})^2 $

$SS_{total} = \sum (y - \bar{y})^2 $

$R^2 = 1- \dfrac{SS_{residual}}{SS_{total}}$

$R^2_{adj}= 1-(1-R^2)\dfrac{n-1}{n-p-1}$

In [39]:
X[selected_cols].shape

(506, 5)

In [41]:
# Your code here
SS_res = np.sum((y-yhat)**2)
SS_tot = np.sum((y-np.mean(y))**2)

r2 = 1 - SS_res/SS_tot
print("R sq = ", r2)

r2_adj = 1 - (1 - r2)* (len(y) - 1)/(len(y)-X[selected_cols].shape[1]-1)
print("R sq adj = ", r2_adj)

# r_squared is 0.742981  
# adjusted_r_squared is 0.740411

R sq =  price    0.742981
dtype: float64
R sq adj =  price    0.740411
dtype: float64


## Level up (Optional)

- Perform variable selection using forward selection, using this resource: https://planspace.org/20150423-forward_selection_with_statsmodels/. Note that this time features are added based on the adjusted R-squared!
- Tweak the code in the `stepwise_selection()` function written above to just perform forward selection based on the p-value 

In [45]:
import statsmodels.formula.api as smf

In [48]:
data = pd.concat([X, y], axis = 1)

In [49]:
data.head()

CRIM     INDUS  CHAS     RM       AGE       DIS   PTRATIO         B  \
0  0.000000 -1.704344   0.0  6.575 -0.120013  0.542096 -1.443977  1.000000   
1  0.153211 -0.263239   0.0  6.421  0.367166  0.623954 -0.230278  1.000000   
2  0.153134 -0.263239   0.0  7.185 -0.265812  0.623954 -0.230278  0.989737   
3  0.171005 -1.778965   0.0  6.998 -0.809889  0.707895  0.165279  0.994276   
4  0.250315 -1.778965   0.0  7.147 -0.511180  0.707895  0.165279  1.000000   

      LSTAT  RAD_(6, 24]  TAX_(270, 360]  TAX_(360, 712]  price  
0 -1.275260            0               1               0   24.0  
1 -0.263711            0               0               0   21.6  
2 -1.627858            0               0               0   34.7  
3 -2.153192            0               0               0   33.4  
4 -1.162114            0               0               0   36.2

In [59]:
from mlxtend.feature_selection import SequentialFeatureSelector as SFS

sfs = sfs = SFS(scikit_model, 
          k_features=5, 
          forward=True, 
          floating=False, 
          scoring='r2',
          cv=4,
          n_jobs=-1)


In [61]:
sfs1 = sfs.fit(X,y)

In [62]:
sfs1.subsets_

{1: {'feature_idx': (8,),
  'cv_scores': array([0.41454506, 0.62677812, 0.42912055, 0.35566007]),
  'avg_score': 0.4565259492939738,
  'feature_names': ('LSTAT',)},
 2: {'feature_idx': (6, 8),
  'cv_scores': array([0.53088579, 0.65409975, 0.50275365, 0.49975095]),
  'avg_score': 0.5468725388013658,
  'feature_names': ('PTRATIO', 'LSTAT')},
 3: {'feature_idx': (6, 7, 8),
  'cv_scores': array([0.53061097, 0.65880814, 0.49763177, 0.51445832]),
  'avg_score': 0.5503772987772443,
  'feature_names': ('PTRATIO', 'B', 'LSTAT')},
 4: {'feature_idx': (2, 6, 7, 8),
  'cv_scores': array([0.55881103, 0.59353947, 0.51757499, 0.49949675]),
  'avg_score': 0.5423555592563075,
  'feature_names': ('CHAS', 'PTRATIO', 'B', 'LSTAT')},
 5: {'feature_idx': (2, 6, 7, 8, 10),
  'cv_scores': array([0.52568807, 0.6035864 , 0.53307399, 0.4901245 ]),
  'avg_score': 0.5381182381411184,
  'feature_names': ('CHAS', 'PTRATIO', 'B', 'LSTAT', 'TAX_(270, 360]')}}

## Summary
Great! You practiced your feature selection skills by applying stepwise selection and recursive feature elimination to the Boston Housing dataset! 